subsys/bluetooth/host/l2cap.c

```diff
@@ -1278,6 +1278,16 @@ static void le_ecred_reconf_req(struct bt_l2cap *l2cap, uint8_t ident,
 			goto response;
 		}
 
+		/* As per BT Core Spec V5.2 Vol. 3, Part A, section 7.11
+		 * The request (...) shall not decrease the MPS of a channel
+		 * if more than one channel is specified.
+		 */
+		if (buf->len >= sizeof(scid) &&
+		    BT_L2CAP_LE_CHAN(chan)->tx.mps > mps) {
+			result = BT_L2CAP_RECONF_INVALID_MPS;
+			goto response;
+		}
```

Comment:
> The request (...) shall not decrease the MPS of a channel if *more than one* channel is specified.

The check `buf->len >= sizeof(scid)` is also true if it is a single channel, right? Shouldn't that be `buf->len > sizeof(scid)` as it is allowed to decrease the MPS if only a single channel is specified.

Reply:
As far as I understand, we are after one `net_buf_pull_le16`, so if `buf->len` still has at least 1 CID it means we got at least 2 to start with

Reply:
Oh, my bad for not checking the previous lines of the function.

I assume that we have a check earlier for valid length as well, so that `buf->len` can never be `1` here, right?

Reply:
I had also missed that this was in a `while` loop. 

Does all the channels always have the same MPS? 
Otherwise, what about the scenario where the first 3 channels updates the MPS (because they had lower MPSs), and then the 4th channel fails this check? Then we should have updated the MTU and MPS of the first 3 channels, and then returned an error to the remove device. 

Is this intentional behavior?

Reply:
You're right, I haven't thought about the case where MPS decreases later than for the first channel. Spec says: 

> In Enhanced Credit Based Flow Control Mode, each individual channel has its
own MTU and MPS; both of these values can be reconfigured independently.

and also. 

> The request shall
not decrease the MTU of any channel and shall not decrease the MPS of a
channel if more than one channel is specified.

So we need to check all requested channels against MPS decrease, and if even for one of them it's lower return the error. I think we might need to save all CIDs first from the request, than loop through them once to check if MPS is always higher, and then loop second time (in `while` loop that's already there) to assign new values, if error was not returned earlier. I don't like this multiple loops and CID saving, but that might be what we need.

Reply:
Yes, it sounds we need a "validate request" step and a "perform request" step, which unfortunately will require two loops.

Reply:
I pushed the patch, sadly, in PTS is a bug that prevents me from testing valid responses (CASE0071492). Tests   L2CAP/ECFC/BI-03-C and L2CAP/ECFC/BI-04-C are passing.